### DISCUSSION TITLE
Migrating App.xaml.cs OnX() to Prism MAUI?

### DISCUSSION THREAD

**Discussion by phillippschmedt** [PrismLibrary/Prism#3036] [.NET MAUI] [answered]
Hi there,

I have been using the App.xaml.cs 

 _protected override void OnStart() 
 protected override void OnResume() 
 protected override void OnSleep()_ 
for Prism/Xamarin.

I notice that in MAUI/Prism we still have the App.xaml.cs. But we also have the PrismBuilder OnAppStart()

How do they compare? When to use the App.xaml.cs OnStart() and when to use the OnAppStart()?

We will be a paying customer starting in January. 

Thanks a lot!

**Comment by dansiegel** [comment, marked as answer]
To be honest `OnAppStart` should probably be renamed `CreateDefaultWindow`. Something to consider when migrating from Xamarin.Forms to .NET MAUI is that while Application.MainPage does still technically exist it is really there for backwards compatibility. Ultimately they should have removed it from the MAUI API.

As for the Application `OnStart`, `OnSleep`, and `OnResume` there are a couple of options you would have. Generally speaking I would suggest using standard Dependency Injection for any services you may need to access in those methods or any others in the Application. However if you needed to resolve the INavigationService I would generally suggest lazy loading it through the `Handler.MauiContext.Services` as this will allow the Application to be initialized and prevent a circular reference.

```cs
public partial class App
{
    private IEventAggregator _ea { get; }
    public App(IEventAggregator ea)
    {
        _ea = ea;
        InitializeComponent();
    }

    protected override void OnStart()
    {
        // invoke an event with the IEventAggregator
    }

    protected override void OnResume()
    {
        var services = Handler.MauiContext.Services;
        if (services is null)
            return;

        var nav = services.GetService<INavigationService>();
        nav.NavigateAsync("/LoginPage");
    }
}
```
